ssm_jax/nlgssm/extended_inference.py

```diff
@@ -0,0 +1,165 @@
+import chex
+import jax.numpy as jnp
+import jax.random as jr
+from jax import lax
+from jax import jacfwd
+from distrax import MultivariateNormalFullCovariance as MVN
+from typing import Callable
+
+@chex.dataclass
+class ESSMParams:
+    """Lightweight container for ESSM parameters.
+    The functions below can be called with an instance of this class.
+    However, they can also accept a ssm.nlgssm.models.ESSM instance,
+    if you prefer a more object-oriented approach.
+    """
+    initial_mean: chex.Array
+    initial_covariance: chex.Array
+    dynamics_function: Callable
+    dynamics_covariance: chex.Array
+    emission_function: Callable
+    emission_covariance: chex.Array
+
+@chex.dataclass
+class ESSMPosterior:
```

Comment:
Maybe call this `NLGSSMPosterior`? (It's actually the same as `LGSSMPosterior` :)
Also this should probably live in the nlgssm/models.py file, since it will be shared between EKF, UKF, etc.

Reply:
Interesting. I wonder if we should use the code to reinforce the point that the EKF/UKF also return the same type of posterior as the Kalman smoother? I.e. they're all linear Gaussian chains. The same is true of Laplace approximations to, e.g., Poisson LDS posteriors.